lightning/src/onion_message/messenger.rs

```diff
@@ -246,6 +246,16 @@ pub trait CustomOnionMessageHandler {
 	fn read_custom_message<R: io::Read>(&self, message_type: u64, buffer: &mut R) -> Result<Option<Self::CustomMessage>, msgs::DecodeError>;
 }
 
+/// A processed incoming onion message, containing either a Forward (another onion message)
+/// or a Receive payload with decrypted contents.
+pub enum PeeledOnion<CMH: Deref> where
```

Comment:
I think it would actually make more sense to just bound `PeeledOnion` by `CustomOnionMessageContents` rather than by the handler trait, since on its own it doesn't need to know about the handler:
```diff
diff --git a/lightning/src/onion_message/messenger.rs b/lightning/src/onion_message/messenger.rs
index 7142e3a5c..1a0145441 100644
--- a/lightning/src/onion_message/messenger.rs
+++ b/lightning/src/onion_message/messenger.rs
@@ -248,13 +248,11 @@ pub trait CustomOnionMessageHandler {

 /// A processed incoming onion message, containing either a Forward (another onion message)
 /// or a Receive payload with decrypted contents.
-pub enum PeeledOnion<CMH: Deref> where
-       CMH::Target: CustomOnionMessageHandler,
-{
+pub enum PeeledOnion<CM: CustomOnionMessageContents> {
        /// Forwarded onion, with the next node id and a new onion
        Forward(PublicKey, msgs::OnionMessage),
        /// Received onion message, with decrypted contents, path_id, and reply path
-       Receive(OnionMessageContents<<<CMH as Deref>::Target as CustomOnionMessageHandler>::CustomMessage>, Option<[u8; 32]>, Option<BlindedPath>)
+       Receive(OnionMessageContents<CM>, Option<[u8; 32]>, Option<BlindedPath>)
 }

 /// Create an onion message with contents `message` to the destination of `path`.
@@ -319,7 +317,7 @@ where
 pub fn peel_onion<NS: Deref, L: Deref, CMH: Deref>(
        node_signer: NS, secp_ctx: &Secp256k1<secp256k1::All>, logger: L, custom_handler: CMH,
        msg: &msgs::OnionMessage,
-) -> Result<PeeledOnion<CMH>, ()>
+) -> Result<PeeledOnion<<<CMH>::Target as CustomOnionMessageHandler>::CustomMessage>, ()>
 where
        NS::Target: NodeSigner,
        L::Target: Logger,

Reply:
for sure, makes more sense this way